docs/why-and-how-of-inline-directives.md

```diff
@@ -0,0 +1,369 @@
+
+```purs
+foo = do
+  let
+    a = { bar: 42, baz: "something" }
+    b = 1 + f a
+  b
+```
+
+**To summarize, a data producer followed immediately by a data consumer indicates a place where an unneeded intermediate data structure exists and can be removed.**
+
+### Removing Unneeded Data Structures
+
+#### Inlining and Inline Directives
+
+An inliner replaces a function's call site (e.g. stuff on the left-hand-side of the `=`) with its implementation (e.g. stuff on the right-hand-side of the `=`). This replacement means the function's body is duplicated and will appear at least twice in the resulting code: once in its original definition and once in the usage site.
+
+Here's one example using the function `binaryPlus`:
+
+```purs
+binaryPlus :: Int -> Int -> Int
+binaryPlus a b = a + b
+
+usage = 9 + binaryPlus 1 2
+```
+
+There's a few ways we could inline the expression, `binaryPlus 1 2`. While `purs-backend-es` will performance its own analysis to determine whether and when to inline it, it doesn't know as much as the developer does.
+
+A developer can use **inline directives** to tell `purs-backend-es` whether to inline a function's body to its usage site and when (i.e. how many args need to be passed to the function before the inlining occurs):
+
+- never inline it at all (i.e. the directive is `never`)
+    - `9 + binaryPlus 1 2`
+- always inline it with no consideration for the number of arguments applied to it (i.e. the directive is `always`)
+    - `9 + (\a b -> a + b) 1 2`
+- inline it only after at least one argument has been applied to it (i.e. the directive is `arity=1`)
+    - `9 + (\  b -> 1 + b)   2`
+- inline it only after at least two arguments have been applied to it (i.e. the directive is `arity=2`)
+    - `9 + (        1 + 2)`
+
+**To summarize, inline directives provide developers with a stronger guarantee about whether and when a function is inlined than just relying upon the analysis done by `purs-backend-es`.**
+
+#### Inlining Duplicates Code
+
+The below example highlights the risk of inlining everything without thought: pointless code duplication.
+
+```purs
+hasExpensiveLetBinding a b = do
+  let
+    someValue = expensiveComputation
+  1 + someValue * b + a
+
+usage1 = hasExpensiveLetBinding 1 2
+usage2 = hasExpensiveLetBinding 3 4
+```
+
+If we inline `hasExpensiveLetBinding` immediately, the `expensiveComputation` will be computed three times rather than once. In other words, it would be the same as writing the following in source code:
+
+```purs
+hasExpensiveLetBinding a b = do
+  let
+    someValue = expensiveComputation
+  1 + someValue * b + a
+
+usage1 = (\a b -> do
+    let
+      someValue = expensiveComputation
+    1 + someValue * b + a
+  ) 1 2
+usage2 = (\a b -> do
+    let
+      someValue = expensiveComputation
+    1 + someValue * b + a
+  ) 3 4
+```
```

Comment:
This example is inaccurate. Inlining here has no effect on how many times `expensiveComputation` is computed, it's always 2 because `expensiveComputation` is deferred under the lambda application.

Reply:
What would be an example that demonstrates this effect while still being correct?

Reply:
Maybe you should just reframe it not in terms of how expensive it is to compute/evaluate, but rather in terms of code size. That is, if `expensiveComputation` (should be renamed) is a large expression that won't be reduced by the optimizer based on usages of `a` or `b`, then you've duplicated code without also reducing runtime overhead.